### PR TITLE
allow the dns check to be ignored with private hosted zones.

### LINKED ISSUE #781: Error doing DNS lookup for NS records when using a Private DNS zone
Since 621dec765da9778b06035a80cdde61216f249934, I have started to get the following error when I try to use KOPS with just a Private DNS zone.

This may impact the current private networking work...

```
AWS_DEFAULT_REGION=us-east-1 AWS_PROFILE=terraform-development KOPS_STATE_STORE=s3://project-infrastructure-development/kops kops create cluster --cloud=aws --associate-public-ip=false --dns-zone MN99W4L99JSA --network-cidr 172.28.0.0/16 --vpc vpc-bbb1k5as --zones=us-east-1a,us-east-1b,us-east-1c,us-east-1e --master-zones=us-east-1a,us-east-1b,us-east-1c  --networking cni --image ami-928a8923 --kubernetes-version v1.4.5 kubernetes.development.project.io
```

```
I1102 14:22:42.100821   93923 cluster.go:421] Assigned CIDR 172.28.32.0/19 to zone us-east-1a
I1102 14:22:42.100861   93923 cluster.go:421] Assigned CIDR 172.28.64.0/19 to zone us-east-1b
I1102 14:22:42.100871   93923 cluster.go:421] Assigned CIDR 172.28.96.0/19 to zone us-east-1c
I1102 14:22:42.100880   93923 cluster.go:421] Assigned CIDR 172.28.160.0/19 to zone us-east-1e
Previewing changes that will be made:

error doing DNS lookup for NS records for "development.project.io": lookup development.project.io on 192.168.1.333:53: no such host
```

Thanks

### PR DESCRIPTION
**Pull request by billyshambrook**
fixes #781

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/810)

### CODE REVIEW

**Comment by billyshambrook** [comment]
I signed it!

**Comment by billyshambrook** [comment]
I have submitted the cla/linuxfoundation CLA.

File: upup/pkg/fi/cloudup/apply_cluster.go [hunk 1]
```diff
@@ -618,24 +619,24 @@ func validateDNS(cluster *api.Cluster, cloud fi.Cloud) error {
 	zone := matches[0]
 	dnsName := strings.TrimSuffix(zone.Name(), ".")
 
-	glog.V(2).Infof("Doing DNS lookup to verify NS records for %q", dnsName)
-	ns, err := net.LookupNS(dnsName)
-	if err != nil {
-		return fmt.Errorf("error doing DNS lookup for NS records for %q: %v", dnsName, err)
-	}
+	if os.Getenv("DNS_IGNORE_NS_CHECK") == "" {
```
Comment (chrislovecnm) [inline upup/pkg/fi/cloudup/apply_cluster.go]: Do we want an env varible or to have a yaml option?

Comment (billyshambrook) [inline upup/pkg/fi/cloudup/apply_cluster.go]: I will be using this env as I am using a private hosted zone. I am unsure if there were other cases when this env var would be used, @justinb? 

Maybe we need to be more explicit being a private hosted zone. We could either identify whether the user has given a private or public using the aws API, or have it explicitly defined in the yaml.

**Comment by billyshambrook** [comment]
OK no worries. So I should create a new yaml option called  `dnsIgnoreNSCheck` which will be used in place of this envvar?

**Comment by billyshambrook** [comment]
The DNS lookup is performed when creating a cluster, so will need to also add a CLI option along with the YAML option.

**Comment by chrislovecnm** [comment]
Performed in the create or update? We can ask the user to do the three step process, create, edit, update.

On the fence. Your call but we already have a ton of cli switches.

**Comment by billyshambrook** [comment]
Sorry, in the `kops create` command - will add the cli switch.

**Comment by jkemp101** [comment]
+1 for this change. I just had to comment out this section of code yesterday to workaround building with a private zone.

**Comment by billyshambrook** [comment]
@chrislovecnm I have replaced the env var with a cli switch (and in the yaml)

**Comment by chrislovecnm** [comment]
@justinsb / @kris-nova I need a model review.  This looks good, but I want it awesome :)

@billyshambrook I am sorry I did not mention this, but what are your thoughts on testing this, or helping improve how we test this.

**Comment by chrislovecnm** [comment]
@billyshambrook great talking with you about test cases.  Like I said don't want 16 million lines of unit tests.  Keep it simple, clean and elegant.

**Comment by chrislovecnm** [comment]
We need a rebase, and I can merge.

File: upup/pkg/fi/cloudup/apply_cluster.go [hunk 2]
```diff
@@ -618,24 +619,24 @@ func validateDNS(cluster *api.Cluster, cloud fi.Cloud) error {
 	zone := matches[0]
 	dnsName := strings.TrimSuffix(zone.Name(), ".")
 
-	glog.V(2).Infof("Doing DNS lookup to verify NS records for %q", dnsName)
-	ns, err := net.LookupNS(dnsName)
-	if err != nil {
-		return fmt.Errorf("error doing DNS lookup for NS records for %q: %v", dnsName, err)
-	}
+	if !fi.BoolValue(cluster.Spec.IgnoreNSCheck) {
```
Comment (justinsb) [inline upup/pkg/fi/cloudup/apply_cluster.go]: This is good, but it doesn't belong in the cluster spec.  Yes, this does mean you'll have to pass the flag every time.

A nicer option is probably explicit support for a private DNS HostedZone.  That didn't make sense before, but now we have topology=private, it does.

**Comment by chrislovecnm** [comment]
We need a rebase and @justinsb has requested changes.

**Comment by chrislovecnm** [comment]
I appreciate your work with this!  Can we get a rebase, and @justinsb has some feedback.

**Comment by chrislovecnm** [comment]
@billyshambrook any updates?

**Comment by chrislovecnm** [comment]
Did we fix this?

**Comment by kris-nova** [comment]
I think so! Adding the scheduled to close label.. Anyone have anything here?

**Comment by kris-nova** [comment]
We haven't seen an update here in some time. Please re open this PR if we still want to follow it through.

**Pull request closed without merging**